src/ResourceStore.php

```diff
@@ -0,0 +1,286 @@
+<?php
+
+namespace silverorange\JsonApiClient;
+
+use GuzzleHttp\Psr7\Request;
+use GuzzleHttp\Client as HttpClient;
+
+class ResourceStore
+{
+    // {{{ protected properties
+
+    protected $json_api_base = null;
+    protected $json_api_headers = null;
+
+    protected $http_client = null;
+
+    protected $class_by_type = array();
+
+    protected $resources = array();
+
+    // }}}
+    // {{{ public function __construct()
+
+    public function __construct($json_api_base, array $json_api_headers = [])
+    {
+        $this->json_api_base = $json_api_base;
+        $this->json_api_headers = $json_api_headers;
+
+        $this->initHttpClient();
+    }
+
+    // }}}
+    // {{{ protected function initHttpClient()
+
+    protected function initHttpClient()
+    {
+        $default_headers = $this->getDefaultHttpHeaders();
+
+        $this->http_client = new HttpClient(
+            [
+                'base_uri' => $this->json_api_base,
+                'headers' => array_merge(
+                    $default_headers,
+                    $this->json_api_headers
+                ),
+            ]
+        );
+    }
+
+    // }}}
+    // {{{ protected function getDefaultHttpHeaders()
+
+    protected function getDefaultHttpHeaders()
+    {
+        return [
+            'Accept' => 'application/vnd.api+json',
+            'Content-Type' => 'application/vnd.api+json',
+        ];
+    }
+
+    // }}}
+    // {{{ public function findAll()
+
+    public function findAll($type, array $query_params = [])
+    {
+        $result = $this->http_client->request(
+            'GET',
+            $this->getResourceAddress($type),
+            ['query' => $query_params]
+        );
+
+        $body = json_decode($result->getBody(), true);
+
+        $collection = new ResourceCollection($type);
+        $collection->setStore($this);
+        $collection->decode($body['data']);
+
+        foreach ($collection as $resource) {
+            $this->setResource(
+                $resource->getType(),
+                $resource->getId(),
+                $resource
+            );
+        }
+
+        return $collection;
+    }
+
+    // }}}
+    // {{{ public function find()
+
+    public function find($type, $id, array $query_params = [])
+    {
+        $resource = $this->peek($type, $id);
+
+        if (!$resource instanceof Resource) {
+            $resource = $this->query($type, $id, $query_params);
+        }
+
+        return $resource;
+    }
+
+    // }}}
+    // {{{ public function query()
+
+    public function query($type, $id, array $query_params = [])
+    {
+        $result = $this->http_client->request(
+            'GET',
+            $this->getResourceAddress($type, $id),
+            ['query' => $query_params]
+        );
+
+        $body = json_decode($result->getBody(), true);
+
+        if (!isset($body['data'])) {
```

Comment:
This should check `if (!is_array($body) || ...) { ...` to prevent trying to access `null['data']` in the event the JSON is non-conformant as well as invalid.